node/external/transactionAPI/apiTransactionProcessor.go

```diff
@@ -351,35 +350,17 @@ func (atp *apiTransactionProcessor) extractRequestedTxInfo(wrappedTx *txcache.Wr
 	return tx
 }
 
-func (atp *apiTransactionProcessor) getDataStoresForSender(senderShard uint32) []storage.Cacher {
-	cachers := make([]storage.Cacher, 0)
-	numOfShards := atp.shardCoordinator.NumberOfShards()
-	for shard := uint32(0); shard < numOfShards; shard++ {
-		cacheId := process.ShardCacherIdentifier(senderShard, shard)
-		shardCache := atp.dataPool.Transactions().ShardDataStore(cacheId)
-		cachers = append(cachers, shardCache)
-	}
-
-	cacheId := process.ShardCacherIdentifier(senderShard, common.MetachainShardId)
-	shardCache := atp.dataPool.Transactions().ShardDataStore(cacheId)
-	cachers = append(cachers, shardCache)
-
-	return cachers
-}
-
 func (atp *apiTransactionProcessor) fetchTxsForSender(sender string, senderShard uint32) []*txcache.WrappedTransaction {
-	txsForSender := make([]*txcache.WrappedTransaction, 0)
-	cachers := atp.getDataStoresForSender(senderShard)
-	for _, cache := range cachers {
-		txCache, ok := cache.(*txcache.TxCache)
-		if !ok {
-			continue
-		}
-
-		txs := txCache.GetTransactionsPoolForSender(sender)
-		txsForSender = append(txsForSender, txs...)
+	cacheId := process.ShardCacherIdentifier(senderShard, senderShard)
```

Comment:
so this will work despite the fact that a transaction might be cross shard? On both nodes: sender & destination?

Reply:
yes, as verified with @andreibancioiu and can be seen [here](https://github.com/ElrondNetwork/elrond-go/blob/8e402fa6d7e91e779980122d3798b2bf50892945/dataRetriever/txpool/shardedTxPool.go#L361), no matter the destination shard, there will always be returned the source id.

the problem was observed on a transaction from shard 0 to shard 1, when the same transaction was returned 4 times from a shard 0 observer.

also, @gabi-vuls confirmed that testing was made with cross shard txs.